### PR TITLE
Allow APPSODY_MOUNT_HOME to override with named docker volume

### PR DESCRIPTION
**Pull request by makandre**
Signed-off-by: Andrew Mak <makandre@ca.ibm.com>

This is a proposed solution for #895 to let `APPSODY_MOUNT_HOME` be used to specify a named docker volume.

For example, if a stack has the following mount:

- `-v ~/.m2/repository:/mvn/repository`

Setting `APPSODY_MOUNT_HOME=named-vol` would override the mount to be:

- `-v named-vol1:/mvn/repository` (number incremented for each home mount)

### CODE REVIEW

File: cmd/utils.go [hunk 1]
```diff
@@ -311,7 +311,12 @@ func getVolumeArgs(config *RootCommandConfig) ([]string, error) {
 		var mappedMount string
 		var overridden bool
 		if strings.HasPrefix(mount, "~") {
-			mappedMount = strings.Replace(mount, "~", homeDir, 1)
+			if homeDirOverride != "" && strings.HasSuffix(homeDir, ":") {
```
[the remark this answers is not in the document]
Comment (EnriqueL8) [reply cmd/utils.go]: We also think that we could have an issue if the stack creator uses multiple mounts within the homedir...(e.g `~/.m2` and `~/.p2`)

[the remark this answers is not in the document]
Comment (makandre) [reply cmd/utils.go]: Sure thing, let me see what I can do about your concerns above.

[the remark this answers is not in the document]
Comment (EnriqueL8) [reply cmd/utils.go]: Thanks!

**Comment by makandre** [comment]
@EnriqueL8 I made changes in ecacdc1 based on your comments.

**Comment by makandre** [comment]
Please let me know which release this will go into.  Thanks for review.

**Comment by Kamran64** [comment]
@makandre - I've tested this change on mac and linux. Works fine on linux but I get this error on mac with the open liberty stack. After setting `APPSODY_MOUNT_HOME` to empty again, it works fine.

```
➜  test git:(review-896) ✗ local-appsody run                               
Running development environment...
[Warning] The Appsody CLI will use the latest version of the controller image. This may result in a mismatch or malfunction.
Pulling docker image docker.io/appsody/java-openliberty:0.2
Running command: docker pull docker.io/appsody/java-openliberty:0.2
0.2: Pulling from appsody/java-openliberty
Digest: sha256:335d2d1255c54bc1b999f0b29958af9f2fe184703f2108657312f671002b91bc
Status: Image is up to date for appsody/java-openliberty:0.2
docker.io/appsody/java-openliberty:0.2
Running command: docker run --rm -v appsody-controller-latest:/.appsody appsody/init-controller:latest
Running command: docker run --rm -p 7777:7777 -p 9080:9080 -p 9443:9443 --name test-dev -u 501:20 -e "APPSODY_USER=501" -e "APPSODY_GROUP=20" -v test-vol1:/mvn/repository -v /Users/kamran.shamsiibm.com/Workspaces/go/src/github.com/appsody/appsody/test/.:/project/user-app -v appsody-controller-latest:/.appsody -t --entrypoint /.appsody/appsody-controller docker.io/appsody/java-openliberty:0.2 "--mode=run"
[Container] Running APPSODY_PREP command: /project/run-stack.sh prep
[Container] Installing parent dev.appsody:java-openliberty:0.2.0
[Container] [INFO] Scanning for projects...
[Container] [INFO] 
[Container] [INFO] --------------------< dev.appsody:java-openliberty >--------------------
[Container] [INFO] Building java-openliberty 0.2.0
[Container] [INFO] --------------------------------[ pom ]---------------------------------
[Container] [INFO] 
[Container] [INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (default-cli) @ java-openliberty ---
[Container] [INFO] Skipping Rule Enforcement.
[Container] [INFO] 
[Container] [INFO] --- maven-install-plugin:2.4:install (default-install) @ java-openliberty ---
[Container] [INFO] Installing /project/user-app/../pom.xml to /mvn/repository/dev/appsody/java-openliberty/0.2.0/java-openliberty-0.2.0.pom
[Container] [INFO] ------------------------------------------------------------------------
[Container] [INFO] BUILD FAILURE
[Container] [INFO] ------------------------------------------------------------------------
[Container] [INFO] Total time:  0.827 s
[Container] [INFO] Finished at: 2020-02-20T14:09:30Z
[Container] [INFO] ------------------------------------------------------------------------
[Container] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:2.4:install (default-install) on project java-openliberty: Failed to install artifact dev.appsody:java-openliberty:pom:0.2.0: /mvn/repository/dev/appsody/java-openliberty/0.2.0/java-openliberty-0.2.0.pom (Permission denied) -> [Help 1]
[Container] [ERROR] 
[Container] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[Container] [ERROR] Re-run Maven using the -X switch to enable full debug logging.
[Container] [ERROR] 
[Container] [ERROR] For more information about the errors and possible solutions, please read the following articles:
[Container] [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[Container] Running command:  /project/run-stack.sh run
[Container] + mvn -B -Plocal-dev -DappsDirectory=apps -Ddebug=false -Dmaven.repo.local=/mvn/repository pre-integration-test liberty:dev
[Container] [INFO] Scanning for projects...
[Container] [INFO] Downloading from central: https://repo.maven.apache.org/maven2/dev/appsody/java-openliberty/maven-metadata.xml
[Container] [WARNING] Failed to write tracking file /mvn/repository/dev/appsody/java-openliberty/resolver-status.properties
[Container] java.io.FileNotFoundException: /mvn/repository/dev/appsody/java-openliberty/resolver-status.properties (Permission denied)
[Container]     at java.io.RandomAccessFile.open0 (Native Method)
[Container]     at java.io.RandomAccessFile.open (RandomAccessFile.java:316)
[Container]     at java.io.RandomAccessFile.<init> (RandomAccessFile.java:243)
[Container]     at org.eclipse.aether.internal.impl.TrackingFileManager.update (TrackingFileManager.java:99)
[Container]     at org.eclipse.aether.internal.impl.DefaultUpdateCheckManager.write (DefaultUpdateCheckManager.java:590)
[Container]     at org.eclipse.aether.internal.impl.DefaultUpdateCheckManager.touchMetadata (DefaultUpdateCheckManager.java:557)
[Container]     at org.eclipse.aether.internal.impl.DefaultMetadataResolver$ResolveTask.run (DefaultMetadataResolver.java:603)
[Container]     at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run (RunnableErrorForwarder.java:75)
[Container]     at org.eclipse.aether.internal.impl.DefaultMetadataResolver$1.execute (DefaultMetadataResolver.java:494)
[Container]     at org.eclipse.aether.internal.impl.DefaultMetadataResolver.resolve (DefaultMetadataResolver.java:350)
[Container]     at org.eclipse.aether.internal.impl.DefaultMetadataResolver.resolveMetadata (DefaultMetadataResolver.java:181)
[Container]     at org.apache.maven.repository.internal.DefaultVersionRangeResolver.getVersions (DefaultVersionRangeResolver.java:198)
[Container]     at org.apache.maven.repository.internal.DefaultVersionRangeResolver.resolveVersionRange (DefaultVersionRangeResolver.java:148)
[Container]     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveVersionRange (DefaultRepositorySystem.java:247)
[Container]     at org.apache.maven.project.ProjectModelResolver.resolveModel (ProjectModelResolver.java:217)
[Container]     at org.apache.maven.model.building.DefaultModelBuilder.readParentExternally (DefaultModelBuilder.java:1070)
[Container]     at org.apache.maven.model.building.DefaultModelBuilder.readParent (DefaultModelBuilder.java:846)
[Container]     at org.apache.maven.model.building.DefaultModelBuilder.build (DefaultModelBuilder.java:337)
[Container]     at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:448)
[Container]     at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:414)
[Container]     at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:377)
[Container]     at org.apache.maven.graph.DefaultGraphBuilder.collectProjects (DefaultGraphBuilder.java:414)
[Container]     at org.apache.maven.graph.DefaultGraphBuilder.getProjectsForMavenReactor (DefaultGraphBuilder.java:405)
[Container]     at org.apache.maven.graph.DefaultGraphBuilder.build (DefaultGraphBuilder.java:82)
[Container]     at org.apache.maven.DefaultMaven.buildGraph (DefaultMaven.java:507)
[Container]     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:219)
[Container]     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
[Container]     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
[Container]     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
[Container]     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
[Container]     at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
[Container]     at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
[Container]     at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
[Container]     at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
[Container]     at java.lang.reflect.Method.invoke (Method.java:498)
[Container]     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
[Container]     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
[Container]     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
[Container]     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[Container] [WARNING] Could not transfer metadata dev.appsody:java-openliberty/maven-metadata.xml from/to central (https://repo.maven.apache.org/maven2): /mvn/repository/dev/appsody/java-openliberty/maven-metadata-central.xml.part.lock (Permission denied)
[Container] [INFO] Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-war-plugin/3.2.2/maven-war-plugin-3.2.2.pom
[Container] [WARNING] Failed to create parent directories for tracking file /mvn/repository/org/apache/maven/plugins/maven-war-plugin/3.2.2/maven-war-plugin-3.2.2.pom.lastUpdated
[Container] [WARNING] Failed to retrieve plugin descriptor for org.apache.maven.plugins:maven-war-plugin:3.2.2: Plugin org.apache.maven.plugins:maven-war-plugin:3.2.2 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.apache.maven.plugins:maven-war-plugin:jar:3.2.2
[Container] [INFO] 
[Container] [INFO] ----------< dev.appsody.starter.java-openliberty:starter-app >----------
[Container] [INFO] Building starter-app 1.0-SNAPSHOT
[Container] [INFO] --------------------------------[ war ]---------------------------------
[Container] [INFO] Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-failsafe-plugin/3.0.0-M1/maven-failsafe-plugin-3.0.0-M1.pom
[Container] [WARNING] Failed to create parent directories for tracking file /mvn/repository/org/apache/maven/plugins/maven-failsafe-plugin/3.0.0-M1/maven-failsafe-plugin-3.0.0-M1.pom.lastUpdated
[Container] [INFO] ------------------------------------------------------------------------
[Container] [INFO] BUILD FAILURE
[Container] [INFO] ------------------------------------------------------------------------
[Container] [INFO] Total time:  0.763 s
[Container] [INFO] Finished at: 2020-02-20T14:09:33Z
[Container] [INFO] ------------------------------------------------------------------------
[Container] [ERROR] Plugin org.apache.maven.plugins:maven-failsafe-plugin:3.0.0-M1 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.apache.maven.plugins:maven-failsafe-plugin:jar:3.0.0-M1: Could not transfer artifact org.apache.maven.plugins:maven-failsafe-plugin:pom:3.0.0-M1 from/to central (https://repo.maven.apache.org/maven2): /mvn/repository/org/apache/maven/plugins/maven-failsafe-plugin/3.0.0-M1/maven-failsafe-plugin-3.0.0-M1.pom.part.lock (No such file or directory) -> [Help 1]
[Container] [ERROR] 
[Container] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[Container] [ERROR] Re-run Maven using the -X switch to enable full debug logging.
[Container] [ERROR] 
[Container] [ERROR] For more information about the errors and possible solutions, please read the following articles:
[Container] [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
[Container] + set +x
[Container] The file watcher is not running because no APPSODY_RUN/TEST/DEBUG_ON_CHANGE action was specified or it has been disabled using the --no-watcher flag.
Closing down development environment.
```

**Comment by makandre** [comment]
@EnriqueL8 What a coincidence, I was just investigating this problem with openliberty stack and was about to open an issue on this.  I'll put a reference to it once it's opened.

**Comment by Kamran64** [comment]
@makandre - I haven't tried this change with other stacks but can you only reproduce on open liberty?

**Comment by makandre** [comment]
I didn't run into this problem on Mac while working on this PR, and I think it's because I was testing against openliberty:0.1 stack

It looks like openliberty:0.2 stack start running as non-root and I am suspecting the error writing to `/mvn/repository` is related to that.

**Comment by makandre** [comment]
@EnriqueL8 What user were you running as on Mac?

**Comment by makandre** [comment]
Opened https://github.com/appsody/stacks/issues/678

**Comment by Kamran64** [comment]
@makandre - I'm only seeing errors for stacks that run as a non root user (which right now is open liberty and spring boot). I tested this change with https://github.com/appsody/stacks/pull/686 as well which still failed. I wonder if it's related to https://github.com/moby/moby/issues/2259 where volumes can only be mounted as root. I'm running as an admin user on mac.

**Comment by makandre** [comment]
I know what's happening here...

The `repository` directory have the following ownership and permissions when a named docker volume is mounted (which I think is "correct" in terms of what the stack intended) :
```
drwxr-xr-x 21 java_user java_group 4096 Feb 21 14:34 repository
```

But the stacks also have `APPSODY_USER_RUN_AS_LOCAL=true`, so when I do `appsody run`, it runs as `uid=501` in my case, and thus cannot write to `repository`

It works when mounting my local `~/.m2/repository` because that directory is own by me (`uid=501`)
It will also work with the named docker mount if I do `appsody run` as root (via `sudo`)

**Comment by Kamran64** [comment]
@makandre - Did https://github.com/appsody/stacks/pull/686 fix this for you? I'm still seeing the same error probably for the reasons you mentioned above.

**Comment by makandre** [comment]
@Kamran64 no it did not, I think the problem that https://github.com/appsody/stacks/pull/686 fixed had the same symptom (permissions denied) but was actually a different issue which manifested on Che.

**Comment by neeraj-laad** [comment]
Incase of bind mount (the standard appsody use case), the contents in the stack image `/mvn/repository` get overwritten by `~/.m2` folder owned by host user, and as we run as the local user it can write to that folder. The slight issue in that flow is stack contents are not used and we install stuff again.

In case of a new volume, the contents of the stack (owned by java_user/group) get written to the volume and then the local user fails to write to it.

@makandre Do you create a new volume? I was wondering if you can create the volume and add a file owned by the local user to solve this?

**Comment by scottkurz** [comment]
I think the fix in https://github.com/appsody/stacks/pull/686 only enables uid 1000 to work without a pre-existing, bind-mounted .m2 (with a pre-existing, bind-mounted .m2 you didn't even need the 686 fix).

But there's still a problem for a uid other than 1000 without a bind-mounted .m2.  

I think the resolution is probably an issue for this particular stack (and others like it).   We either have to:

1. loosen the permissions so that "other" (anyone) can write to /mvn/repository
2. Build out the "java_group" support by documenting/clarifying what steps are needed to get access by virtue of group membership (I think this would still need a fix because the /mvn/repository entries I'm seeing are 755, not group-writable)
3. Say this is a uid 1000 stack.. build your own stack differently if you want different behavior.

I can't see that we'd do 3. since it's just broken right off the bat with Mac.. just saying at some level it is an "option".

**Pull request merged**